tools/audio-codec-cli/src/feedback.rs

```diff
@@ -94,3 +99,199 @@ pub async fn feedback(args: StaticArgs) -> anyhow::Result<()> {
+                    input_fell_behind = true;
+                    0.0
+                }
+            };
+        }
+        if input_fell_behind {
+            eprintln!("input stream fell behind: try increasing latency");
+        }
+    };
+
+    // Build streams.
+    println!(
+        "Attempting to build both streams with f32 samples and `{:?}`.",
+        config
+    );
+    let input_stream = input_device.build_input_stream(&config, input_data_fn, err_fn, None)?;
+    let output_stream = output_device.build_output_stream(&config, output_data_fn, err_fn, None)?;
+    println!("Successfully built streams.");
+
+    input_stream.play()?;
+    output_stream.play()?;
+
+    // Run for 3 seconds before closing.
+    println!("Playing for 10 seconds... ");
+    tokio::time::sleep(std::time::Duration::from_secs(10)).await;
+    drop(input_stream);
+    drop(output_stream);
+    println!("Done!");
+    Ok(())
+}
+
+pub async fn echo_cancellation(args: StaticArgs) -> anyhow::Result<()> {
+    let mut input_frame: Vec<f32> = Vec::new();
+    input_frame.reserve(480);
+
+    let mut output_frame: Vec<f32> = Vec::new();
+    output_frame.reserve(480);
+
+    let mut processor = webrtc_audio_processing::Processor::new(&InitializationConfig {
+        num_capture_channels: 1,
+        num_render_channels: 1,
+        ..Default::default()
+    })?;
+    let config = webrtc_audio_processing::Config {
+        echo_cancellation: Some(EchoCancellation {
+            suppression_level: EchoCancellationSuppressionLevel::Moderate,
+            stream_delay_ms: None,
+            enable_delay_agnostic: true,
+            enable_extended_filter: true,
+        }),
+        enable_high_pass_filter: true,
+        ..Default::default()
+    };
+    processor.set_config(config);
+
+    let host = cpal::default_host();
+    let input_device = host.default_input_device().unwrap();
+    let output_device = host.default_output_device().unwrap();
+
+    println!("Using input device: \"{}\"", input_device.name()?);
+    println!("Using output device: \"{}\"", output_device.name()?);
+
+    // We'll try and use the same configuration between streams to keep it simple.
+    let config: cpal::StreamConfig = cpal::StreamConfig {
+        channels: 1,
+        sample_rate: SampleRate(args.sample_rate),
+        buffer_size: cpal::BufferSize::Default,
+    };
+
+    // The buffer to share samples
+    let ring = HeapRb::<f32>::new(48000 * 2);
+    let (mut producer, mut consumer) = ring.split();
+
+    // latency
+    for _ in 0..480 {
+        let _ = producer.push(0.0);
+    }
+
+    let processor = Arc::new(Mutex::new(processor));
+    let processor2 = processor.clone();
+
+    let input_data_fn = move |data: &[f32], _: &cpal::InputCallbackInfo| {
+        let mut input_fell_behind = false;
+        for &sample in data {
+            input_frame.push(sample);
+            if input_frame.len() == 480 {
+                let mut ap = match processor.lock() {
```

Comment:
Is this lock from std or does it reference a result? If the former, you could access the inner value when it errors and use that too, but that may not be safe to do depending on the actual cause of the error.

Reply:
the lock is from std